Optimize DynamicGraph.getUnfinishedNodes

krmahadevan opened the pull request:
Closes #1472

Fixes #1472 .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

juherr commented on src/main/java/org/testng/TestRunner.java:
```diff
@@ -1064,7 +1065,10 @@ private void afterRun() {
-      result.addNode(m);
+      boolean added = result.addNode(m);
+      if (!added) {
+        result.addNode(new WrappedTestNGMethod(m));
```
What if we always use WrappedTestNGMethod?

krmahadevan replied:
> What if we always use WrappedTestNGMethod?

Not sure if its needed. The only time wherein I thought we needed this was when our users used a method interceptor, which internally bumped up the number of test methods (Please refer [here](https://github.com/cbeust/testng/blob/master/src/test/java/test/verify/VerifyTest.java#L22-L24) )

juherr replied:
I think it will help the readability and simplify the code if we have only one generic workflow.
I think some comments will help too.

krmahadevan replied:
I will add comments and push it in once I go home. But we still need the conditional insertion of `WrappedTestNGMethod` because if we dont do it, then we end up with the original situation wherein methods are being lost. Here's a scenario 

Lets say I have 2 methods and a user used a Method interceptor and then changed it to 3 methods by duplicating one of the methods. Now if we wrap all methods within the `WrappedTestNGMethod` then our logic will fail when we are attempting at adding the new method instance which was literally cloned because it would have the same `hashCode` (our implementation just falls back to the hashcode value of `Method` which in this case would be the same). That is why we first need to try injecting the `TestNGMethod` instance. if its injection fails then we try and inject it via `WrappedTestNGMethod` (which internally fudges the hashCode and ensures that it would be different). I am thinking that I should perhaps change that constant `31` and instead use a random number as the multiplication factor, because I am thinking that it will still cause problems if the user attempts duplicating the same method multiple times and then we try creating a `DynamicGraph` out of it.

juherr replied:
Great! I think the test cases you described should be added too ;)

krmahadevan replied:
@juherr - Sure thing. I think that test case itself came from [this](http://beust.com/weblog/2010/03/23/better-mock-testing-with-testng/) blog post from @cbeust (I remember that `@Verify` annotation from his blog only). This time I will add these changes as an additional commit so that its easy for vetting out.

juherr commented on src/main/java/org/testng/internal/DynamicGraph.java:
```diff
@@ -40,11 +40,27 @@ private Edge(int weight, T from, T to) {
+      m_nodesReady = Lists.newArrayList();
+      m_nodesRunning = Lists.newArrayList();
+    } else {
+      m_nodesFinished = Sets.newLinkedHashSet();
```
What if we always use HashSet?

juherr replied:
Where do we use `new DynamicGraph(true)`?

krmahadevan replied:
> What if we always use HashSet?

The implementation is using HashSet only by default.

> Where do we use new DynamicGraph(true)?

We use `new DynamicGraph(true)` only in the test wherein we want to ensure that the time taken for the set implementation is lesser than the time taken for the list backed implementation.

juherr replied:
oops, bad eyes, sorry.

please, remove lists because you already proved that the change is better.

krmahadevan replied:
Sure I will remove off the lists implementation as well.

krmahadevan replied:
@juherr - btw, i just now realised.. If I remove off the lists support, then I will have to get rid of the `performanceTest()` test case as well, because now it wont have anything to assert against. Is that fine ?

juherr replied:
Sure. But if you really want to compare perfs in the test, you'll have to pass collections in the constructor (default constructor with Sets).

juherr commented on src/main/java/org/testng/internal/WrappedTestNGMethod.java:
```diff
@@ -0,0 +1,376 @@
+ * that it wraps.
+ */
+public class WrappedTestNGMethod implements ITestNGMethod {
+    private ITestNGMethod testNGMethod;
```
final

krmahadevan replied:
fixed.

juherr commented on src/main/java/org/testng/internal/WrappedTestNGMethod.java:
```diff
@@ -0,0 +1,376 @@
+    }
+
+    @Override
+    public Class getRealClass() {
```
Class<?>

krmahadevan replied:
fixed.

juherr commented on src/main/java/org/testng/internal/WrappedTestNGMethod.java:
```diff
@@ -0,0 +1,376 @@
+    @Override
+    public void setTestClass(ITestClass cls) {
+        testNGMethod.setTestClass(cls);
+
```
empty line

krmahadevan replied:
fixed.

juherr commented on src/main/java/org/testng/internal/WrappedTestNGMethod.java:
```diff
@@ -0,0 +1,376 @@
+
+    @Override
+    public int hashCode() {
+        return 31* testNGMethod.hashCode();
```
space missing before *

krmahadevan replied:
fixed

krmahadevan commented:
> Looks good but the optimisation is not used and what is the approximative perf optimisation?

Not sure what you mean by its not used. DynamicGraph now uses only the HashSet. 

> what is the approximative perf optimisation?

The naive performance test that I added to `DynamicGraphTest` attempts at creating 1000 objects each of which introduces a lag of 1 to 50 ms when attempting to compute `equals()`.

Below are the results in `ms` for a couple of iterations

```
Time taken for list backed implementation : 11585
Time taken for set backed implementation : 1

Time taken for list backed implementation : 43181
Time taken for set backed implementation : 26

Time taken for list backed implementation : 23456
Time taken for set backed implementation : 42

Time taken for list backed implementation : 11565
Time taken for set backed implementation : 47

Time taken for list backed implementation : 43192
Time taken for set backed implementation : 13


Time taken for list backed implementation : 23320
Time taken for set backed implementation : 1

Time taken for list backed implementation : 11857
Time taken for set backed implementation : 1

Time taken for list backed implementation : 11816
Time taken for set backed implementation : 25
```

krmahadevan commented:
@juherr - I have amended the commit and pushed in the changes to address the review feedback. Please take a look and let me know if its fine.

juherr commented:
The perf improvement is awesome :D

krmahadevan commented:
> The perf improvement is awesome :D

Yeah.. Many thanks to @nathanila. 👍

krmahadevan commented:
ping @juherr - I have addressed the review comments as a separate commit and removed the performance test changes as a separate commit. Please help take a look.
